src/System.Management.Automation/engine/hostifaces/LocalConnection.cs

```diff
@@ -811,48 +811,41 @@ private void CloseThreadProc()
         /// </remarks>
         private void DoCloseHelper()
         {
-            // Stop any transcription if we're the last runspace to exit
-            ExecutionContext executionContext = this.GetExecutionContext;
-            if (executionContext != null)
+            var isPrimaryRunspace = (Runspace.PrimaryRunspace == this);
+            var haveOpenRunspaces = false;
+            foreach (Runspace runspace in RunspaceList)
             {
-                Runspace hostRunspace = null;
-                try
-                {
-                    hostRunspace = executionContext.EngineHostInterface.Runspace;
-                }
-                catch (PSNotImplementedException)
+                if (runspace.RunspaceStateInfo.State == RunspaceState.Opened)
                 {
-                    // EngineHostInterface.Runspace throws PSNotImplementedException if there
-                    // is no interactive host.
+                    haveOpenRunspaces = true;
+                    break;
                 }
+            }
 
-                if ((hostRunspace == null) || (this == hostRunspace))
-                {
-                    // We should close transcripting only if we are closing the last opened runspace.
-                    foreach (Runspace runspace in RunspaceList)
-                    {
-                        // At this stage, the last opened runspace should be at closing state.
-                        if (runspace.RunspaceStateInfo.State == RunspaceState.Opened)
-                        {
-                            return;
-                        }
-                    }
+            // When closing the primary runspace, ensure all other local runspaces are closed.
+            var closeAllOpenRunspaces = isPrimaryRunspace && haveOpenRunspaces;
 
-                    PSHostUserInterface host = executionContext.EngineHostInterface.UI;
-                    if (host != null)
+            // Stop all transcriptions and unitialize AMSI if we're the last runspace to exit or we are exiting the primary runspace.
+            if (isPrimaryRunspace || !haveOpenRunspaces)
```

Comment:
Can you please respond to my previous comment here?
> Perhaps this check should be `if (!haveOpenRunspaces)`. When closing the default runspace, it's possible there are other runspaces still running code, which may be calling into `AmsiUtils`, so calling `AmsiUtils.Uninitialize()` may cause unexpected error.
`AmsiUtils.Uninitialize()` and `StopAllTranscribing()` will called anyway when the last runspace is being closed.

Reply:
I agree.  Will just use haveOpenRunspaces.